### PR TITLE
[stdlib] Follow up improvements to stdlib process module

### PR DESCRIPTION
**Pull request by izo0x90**
This is a follow up this PR https://github.com/modular/modular/pull/3998/changes, which introduces the process module to stdlib.

The following improvements (discussed in the original PR) are being added:
- Stdio redirection through file descriptors (now the Pipe object can be used for comm. with sub process)
- Passing a mapping of env. variables for sub process
- Structs for managing Args, Env, FileActions to allow for RAII style resource allocation (to not required complex deallocation logic for edge cases in process.run) 
- Further test coverage for new and existing module functionality

### CODE REVIEW

**Comment by izo0x90** [comment]
@barcharcraz Hey Charlie, here are the follow ups from the additional commits to the initial process module PR. This hopefully gets the module to state where it covers most basic/ simple use cases. Let me know what you think once you get a chance to look, thank you!

CC: @JoeLoser

File: mojo/stdlib/std/os/process.mojo [hunk 1]
```diff
@@ -107,8 +113,8 @@ struct Pipe:
 
     fn __init__(
         out self,
-        in_close_on_exec: Bool = True,
-        out_close_on_exec: Bool = True,
+        in_close_on_exec: Bool = False,
+        out_close_on_exec: Bool = False,
```
Comment (barcharcraz) [inline mojo/stdlib/std/os/process.mojo:117]: why did these change?

File: mojo/stdlib/std/os/process.mojo [hunk 2]
```diff
@@ -183,7 +189,7 @@ struct Pipe:
             raise Error("Can not write from read only side of pipe")
 
     @always_inline
-    fn read_bytes(mut self, buffer: Span[mut=True, Byte]) raises -> UInt:
+    fn read_bytes(mut self, mut buffer: Span[mut=True, Byte]) raises -> UInt:
```
Comment (barcharcraz) [inline mojo/stdlib/std/os/process.mojo:192]: why do we need this to be mut? The span's length is not modified.

File: mojo/stdlib/std/sys/_libc.mojo [hunk 3]
```diff
@@ -105,13 +106,100 @@ struct BufferMode:
...
+    Opaque, platform-dependent structure used to define file actions
+    for `posix_spawnp`. Its internal layout varies by operating system's
+    C library implementation (e.g., glibc on Linux, macOS's libc).
+
+    This definition uses a fixed-size opaque buffer (128 bytes) to
+    accommodate the largest known common implementations (e.g., glibc
+    on x86-64 requires approx. 80 bytes; macOS uses a pointer-to-pointer
+    model which also fits within this).
+
+    Slightly over sized to ensure future compatibility.
```
Comment (barcharcraz) [inline mojo/stdlib/std/sys/_libc.mojo:120]: we could expand it later if we ever support a platform that needs more space.

File: mojo/stdlib/std/os/process.mojo [hunk 4]
```diff
@@ -201,35 +207,163 @@ struct Pipe:
...
+        UnsafePointer[mut=False, c_char, ImmutAnyOrigin], MutAnyOrigin
+    ]:
+        return self._ptr
+
+
+struct _FileActions:
+    """RAII wrapper for posix_spawn_file_actions_t."""
+
+    var _actions: posix_spawn_file_actions_t
+    var _initialized: Bool
```
Comment (barcharcraz) [inline mojo/stdlib/std/os/process.mojo:293]: could be Optional[posix_spawn_file_actions_t]

Comment (barcharcraz) [reply mojo/stdlib/std/os/process.mojo:293]: actually, we could just always initialize it and have the spawn wrapper take an Optional[_FileActions]

File: mojo/stdlib/std/os/process.mojo [hunk 5]
```diff
@@ -364,61 +508,60 @@ struct Process:
...
+        argv: List[String],
+        env: Optional[Dict[String, String]] = None,
+        stdin: Optional[FileDescriptor] = None,
+        stdout: Optional[FileDescriptor] = None,
+        stderr: Optional[FileDescriptor] = None,
+    ) raises -> Process:
         """Spawn new process from file executable.
 
         Args:
           path: The path to the file.
```
Comment (barcharcraz) [inline mojo/stdlib/std/os/process.mojo:522]: should probably note  that we're calling spawnp and thus search PATH

File: mojo/stdlib/test/os/test_process_io.mojo [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    var result_span = buf_span.unsafe_subspan(offset=0, length=Int(bytes_read))
+    var result = String(bytes=result_span)
+    assert_equal(result, "hello")
+    buf_ptr.free()
+
+
+def main():
+    test_stdout_redirect()
+    test_stderr_redirect()
+    test_stdin_redirect()
```
Comment (barcharcraz) [inline mojo/stdlib/test/os/test_process_io.mojo:84]: can use `TestSuite.discover_tests[__functions_in_module()]().run()`

File: mojo/stdlib/test/os/test_process_logic.mojo [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    assert_equal(status.term_signal.value(), SignalCodes.KILL)
+
+
+def main():
+    test_exit_code_verification()
+    test_process_exits_immediately()
+    test_multiple_wait_calls()
+    test_signal_hangup()
+    test_signal_interrupt()
+    test_signal_kill()
```
Comment (barcharcraz) [inline mojo/stdlib/test/os/test_process_logic.mojo:87]: ditto on `TestSuite.discover_tests[__functions_in_module()]().run()`

**Review by barcharcraz — CHANGES_REQUESTED** [review]
